### PR TITLE
Revamp analytics

### PR DESCRIPTION
**Pull request by irevoire**
# Pull Request

## Related issue
Fixes https://github.com/meilisearch/meilisearch/issues/5009

## What does this PR do?
- Force every analytics to go through a trait that forces you to handle aggregation correcty
- Put the code to retrieve the `user-agent`, `timestamp` and `requests.total_received` in common between all aggregates, so there is no mistake
- Get rids of all the different channel for each kind of event in favor of an any map
- Ensure that we never [send empty event ever again](https://github.com/meilisearch/meilisearch/pull/5001)
- Merge all the sub-settings route into a global « Settings Updated » event.
- Fix: When using one of the three following feature, we were not sending any analytics IF they were set from the global route
  - /non-separator-tokens
  - /separator-tokens
  - /dictionary

### CODE REVIEW

File: meilisearch/src/routes/indexes/documents.rs [hunk 1]
```diff
@@ -100,12 +102,89 @@ pub struct GetDocument {
...
 
+aggregate_methods!(
+    DocumentsGET => "Documents Fetched GET",
+    DocumentsPOST => "Documents Fetched POST",
+);
+
+#[derive(Default, Serialize)]
+pub struct DocumentsFetchAggregator<Method: AggregateMethod> {
+    #[serde(rename = "requests.total_received")]
+    total_received: usize,
```
Comment (ManyTheFish) [inline meilisearch/src/routes/indexes/documents.rs:113]: default of `usize` is `0` so aggregation may always return `0`

File: meilisearch/src/routes/indexes/documents.rs [hunk 2]
```diff
@@ -100,12 +102,89 @@ pub struct GetDocument {
     retrieve_vectors: Param<bool>,
 }
 
+aggregate_methods!(
+    DocumentsGET => "Documents Fetched GET",
+    DocumentsPOST => "Documents Fetched POST",
+);
+
+#[derive(Default, Serialize)]
```
Comment (ManyTheFish) [inline meilisearch/src/routes/indexes/documents.rs:110]: We may prefer avoid the `Default` impl, too error prone.
```suggestion
#[derive(Serialize)]
```

File: meilisearch/src/routes/indexes/documents.rs [hunk 3]
```diff
@@ -129,17 +211,54 @@ pub async fn get_document(
     Ok(HttpResponse::Ok().json(document))
 }
 
+#[derive(Default, Serialize)]
+pub struct DocumentsDeletionAggregator {
+    #[serde(rename = "requests.total_received")]
+    total_received: usize,
```
Comment (ManyTheFish) [inline meilisearch/src/routes/indexes/documents.rs:217]: Is `total_received` necessary here?

File: meilisearch/src/routes/indexes/documents.rs [hunk 4]
```diff
@@ -298,23 +421,72 @@ fn from_char_csv_delimiter(
     }
 }
 
+aggregate_methods!(
+    Replaced => "Documents Added",
+    Updated => "Documents Updated",
+);
+
+#[derive(Default, Serialize)]
```
Comment (ManyTheFish) [inline meilisearch/src/routes/indexes/documents.rs:429]:
```suggestion
#[derive(Serialize)]
```

File: meilisearch/src/routes/indexes/documents.rs [hunk 5]
```diff
@@ -129,17 +211,54 @@ pub async fn get_document(
     Ok(HttpResponse::Ok().json(document))
 }
 
+#[derive(Default, Serialize)]
```
Comment (ManyTheFish) [inline meilisearch/src/routes/indexes/documents.rs:214]:
```suggestion
#[derive(Serialize)]
```

File: meilisearch/src/routes/indexes/documents.rs [hunk 6]
```diff
@@ -599,13 +793,41 @@ pub struct DocumentEditionByFunction {
     pub function: String,
 }
 
+#[derive(Default, Serialize)]
```
Comment (ManyTheFish) [inline meilisearch/src/routes/indexes/documents.rs:796]:
```suggestion
#[derive(Serialize)]
```

**Review by ManyTheFish — CHANGES_REQUESTED** [review]
Some changes before the WE, sorry :/

**Comment by irevoire** [comment]
I think I messed up the settings. When I started using the or instead of the map_or stuff, I should have swapped the order of the self and other field to select the new one instead of the old one.

+ I want to rename the argument of the aggregate function trait to call it "new" instead of "other" so it's clearer which version is new and old

**Comment by irevoire** [comment]
I don’t think we need to integrate the helper function into the trait, we can simply specialize it at compile time and get a ptr to it without putting it in the trait

**Pull request merged**